### ISSUE TITLE
Invalid marks in schema are not removed

### ISSUE THREAD

**Issue by PierBover** [ianstormtaylor/slate#1792]
#### Bug report

With this schema it shouldn't be possible to have any marks in the `heading-one` block:
```
const schema = {
  blocks: {
    'heading-one': {
      nodes: [{ objects: ['text'] }],
      marks: []
    }
  }
};
```
But as you can see I'm able to insert bold text in it:
![image](https://user-images.githubusercontent.com/226483/39166782-2aca079a-4751-11e8-98d5-b989999c9889.png)

#### What's the current behavior?
You can see this behaviour in action here:
https://youtu.be/mfI3U3KX9ug

To reproduce I'm just toggling a mark without actually making a selection and then simply writing.

**Comment by CameronAckermanSEL** [comment]
Try `marks: ['']`

**Comment by CameronAckermanSEL** [comment]
Also there's a [slack channel](https://slate-js.slack.com/messages/C1RH7AXSS/), which is a better place to ask support questions like this:

**Comment by PierBover** [comment]
Thanks @CameronAckermanSEL but I'm getting the same result with either `[]` or `['']`.

**Comment by CameronAckermanSEL** [comment]
The `['']` should definitely work so there's something else in play. 

Should probably go to the slack channel for some support.

**Comment by CameronAckermanSEL** [comment]
Sorry, misunderstood. Able to reproduce this bug, specifically when inserting text. Appears to be because the insertion of text doesn't cause normalization so it doesn't pick up schema violations like this anymore.

**Comment by PierBover** [comment]
Is this a good first task? I'd like to contribute to Slate.

**Comment by zhujinxuan** [comment]
@PierBover Yes, you can fix it here: packages/slate/src/changes/by-key in insertTextAtRange
```
  if (normalize !== undefined) {
    normalize = range.isExpanded
  }
```
to something like
```
  if (normalize !== undefined) {
    normalize = range.isExpanded && marks.size !== 0
  }
```

**Comment by PierBover** [comment]
Ok I'll take a look tonight.

**Comment by SmilinBrian** [comment]
A discussion with "bupkis" on Slack on May 29th shows that this issue also applies to cases where the editor is intended to automatically convert certain sequences of characters typed into macros or something.

I made a JSFiddle showing this issue so that this other aspect of this problem can be fixed/tested as well:  https://jsfiddle.net/SmilinBrian/d5qrhtgx/2/

**Comment by zhujinxuan** [comment]
@SmilinBrian Can you move the discussion to https://github.com/ianstormtaylor/slate/pull/1823

**Comment by isubasti** [comment]
@SmilinBrian you might want to use something like https://github.com/ianstormtaylor/slate-plugins/tree/master/packages/slate-auto-replace rather than validateNode for that

**Comment by samuelwagen** [comment]
@isubasti slate-auto-replace is only suitable for when the user actually types. It doesn't work for example if you also need to process the initial state.

**Comment by isubasti** [comment]
@samuelwagen yeah the `schema normalization` is still needed for some circumstances but for the use cases that @SmilinBrian pointed out, which is for macros then it will work. I believe it should work as it is anyway for the schema normalization on initial state

**Comment by CameronAckermanSEL** [comment]
@samuelwagen this issue is in regards to `insertTextAtRange` not normalizing, so I think `slate-auto-replace` would still be a more appropriate approach to achieving the use case that @SmilinBrian described, as both `insertTextAtRange` and `slate-auto-replace` are tied closely to `onKeyDown` events. 

Cases where you would want schema normalization to be applied to the text on load are out of scope for this issue and for `slate-auto-replace`

**Comment by SmilinBrian** [comment]
I agree that normalization on load is out of scope for this issue.

I still feel ambivalent about whether or not `validateNode()` should be called for all text insertion regardless of whether prior selection is collapsed. Seems to me it should be called, assuming validation is properly limited to "only the affected node(s)." If limited, any performance issues are due to the validations being too complex.  That is, I'm assuming the built-in validations are fairly performant, and it would only be custom validations that might cause performance problems.

**Comment by CameronAckermanSEL** [comment]
There are a couple of open PRs related to `insertTextAtRange` that address this. 

The first of these triggers normalization when there's a mark, which I think is one case where normalization needs to be evaluated on insertion of text.

https://github.com/ianstormtaylor/slate/pull/1875
https://github.com/ianstormtaylor/slate/pull/1823
